rfcs/0013-transition-caching.md

```diff
@@ -0,0 +1,40 @@
+# Transition Caching
+
+## Summary
+[summary]: #summary
+
+A new transition caching logic for the transition router which aims to track transitions which are still being processed as well as transitions which have failed validation.
+
+## Motivation
+[motivation]: #motivation
+
+Within the transition router system, the only check for duplicate transitions is performed by the transition validator, and each transition is only checked against the transitions which are currently in the transition frontier. However, there are two types of duplicate transitions which are not being checked for: transitions which are still being processed by the system (either in the processor pipe or in the catchup scheduler and catchup thread), and transitions which have been determined to be invalid. In the case of the former, the system ends up processing more transitions than necessary, and the number of duplicated processing increases along with the networks size. In the case of the latter, the system is opened up for DDoS attacks since an adversary could continously send transitions with valid proofs but invalid staged ledger diffs, causing each node to spend a significant enough amount of time before invalidating the transition each time it recieves it.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+The goal here is to introduce 2 new caches to the system: the `Invalid_transition_cache`, and the `Unprocessed_transition_cache`.
+
+`Invalid_transition_cache` would be a bloom filter based cache which is updated whenever any form of validation fails on a transition. A bloom filter is a good fit since it remains a constant size while giving a probablistic check to see if an item is in the set or not. A bloom filter can produce false negatives for the existence of an item in the set, but this is considered ok since we are still performing validation on the transition after the `Invalid_transition_cache` is check. The check for this cache would be hoisted up to the toplevel `Transition_router` layer of the transition handling system, as this check is equally shared across both `Bootstrap_controller` and `Transition_frontier_controller`. The `Invalid_transition_cache` should also be persisted to disk. Depending on the cost of writing it to disk, it will either be written out on every update, or written out on an interval.
```

Comment:
I thought you do not want Validator to know about the transitions in catchup_scheduler. What changed your mind?

Reply:
What I didn't want to do is have the validator manage the state of whether or not transitions should go to bootstrap, which requires looking up parent hashes in the proposer pipe. This change still does not do that, it only increases the scope of the validator's check to see if a transition has been seen before (that check already exists, but it only checks the transition frontier's current state).

Reply:
Can you note that in the future we would extend this bloom filter to cache all invalid messages not just transitions.

Reply:
You have the behavior of bloom filters backwards. You can have false positives but not false negatives. If an item is added to a bloom filter, the hashes are computed and all the corresponding bits are set, so lookup will always return true. But the corresponding bits for a given item may be set coincidentally by the insertion of other items, so lookup may return true for items that were not inserted.